## Prediction of CO2 emissions from country-specific data 

A Machine Learning project
---

***
## Stage 1: Data cleaning and preparation
***

## Notebook Contents:
0. Introduction - project and notebook summary, notes on the data source
1. Notebook setup - libraries and data import
2. Global data overview
3. Definition of the initial project goals
4. Data cleaning
   - dealing with missing values
   - transformation of the columns into a numerical data type
   - renaming of features
   -  removing empty columns and rows
5. Data frame transformation
   - melting of the data for each variable
   - integration of the data into a suitable data frame format
6. Removal of missing values
   - detection of missing values
   - removal of missing values by filtering the columns and rows, so that        minimal amount of features and rows are lost
7. Export the clean data frame to a file

#  0. Introduction

##  Project Summary

**Aim of the Project:**  
Analyze **country-specific data** and develop **machine learning models** to **predict CO₂ emissions** from various country parameters.

This project uses the publicly available dataset **Climate Change Data** from the **World Bank Group**, which provides detailed data for most countries over a range of years on parameters such as:

- **Country:** Data covers the majority of countries worldwide  
- **Year:** Ranges from **1990 to 2011**  
- **Greenhouse gas emissions:** CO₂, CH₄, N₂O, and others  
- **Population-related parameters:** Population count, urban population, population growth, etc.  
- **Economic indicators:** GDP, GNI, Foreign Direct Investment, etc.  
- **Land-related parameters:** Cereal yield, agricultural land, protected areas, etc.  
- **Climate data:** Precipitation, natural disasters, etc.  
- **Energy use**  
- **Health indicators:** Counts of medical personnel  
- **Other environmental and social parameters**

---

## Project Structure

The project is divided into **three main stages**:

1. **Data Cleaning and Preparation**  
   - Organizing, transforming, and preparing data for analysis.

2. **Data Exploration and Visualization**  
   - Understanding patterns and relationships through visual analysis.

3. **Predictive Analysis using Random Forest Algorithm**  
   - Building and evaluating machine learning models for prediction.

Each stage is described in a separate **Jupyter Notebook (`.ipynb`)** and a corresponding **PDF file**.

---

## Notebook Summary — Stage 1: Data Cleaning and Preparation

**Aim of this Notebook:**  
Explain and perform the **first stage** of the project — the **cleaning and transformation** of the available data to prepare it for visualization and further analysis.

- **Input:** Excel data file from the original online source  
- **Output:** Cleaned **CSV file** ready for visualization and machine learning  
- **Programming Language:** Python 3.7  
- **Libraries Used:** `pandas`, `numpy`

---

## Data Source

The dataset used in this project is from the **Climate Change Data** provided by the **World Bank Group**.  
It contains **country-specific information** on parameters such as:
- CO₂ emissions  
- Energy use  
- Population count  
- Urban population  
- Cereal yield  
- Nationally protected land areas  
- GDP and GNI  

**Source:** [World Bank Climate Change Data](https://datacatalog.worldbank.org/dataset/climate-change-data)  
**License:** Creative Commons Attribution 4.0 International (CC BY 4.0)

---


## 1. Notebook Setup
Import all needed libraries:

In [2]:
import pandas as pd 
import numpy as np 

In [38]:
# define the file name and the data sheet
orig_data_file = r"C:\Users\yashn\Downloads\climate_change_download_0 (1).xls"
data_sheet = "Data"

# read the data from the excel file to a pandas DataFrame
data_orig = pd.read_excel(io=orig_data_file, sheet_name=data_sheet)

 

---

## 2. Global data overview
A global overview of the imported data yields the following insights:

In [42]:
print("Shape of the original dataset:")
data_orig.shape

Shape of the original dataset:


(13512, 28)

In [44]:
print("Available columns:")
data_orig.columns

Available columns:


Index(['Country code', 'Country name',  'Series code',  'Series name',
              'SCALE',     'Decimals',           1990,           1991,
                 1992,           1993,           1994,           1995,
                 1996,           1997,           1998,           1999,
                 2000,           2001,           2002,           2003,
                 2004,           2005,           2006,           2007,
                 2008,           2009,           2010,           2011],
      dtype='object')

In [46]:
print("Column data types:")
data_orig.dtypes

Column data types:


Country code    object
Country name    object
Series code     object
Series name     object
SCALE           object
Decimals        object
1990            object
1991            object
1992            object
1993            object
1994            object
1995            object
1996            object
1997            object
1998            object
1999            object
2000            object
2001            object
2002            object
2003            object
2004            object
2005            object
2006            object
2007            object
2008            object
2009            object
2010            object
2011            object
dtype: object

In [48]:
print("Overview of the first 5 rows:")
data_orig.head()

Overview of the first 5 rows:


Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   
2          AFG  Afghanistan  AG.LND.EL5M.ZS   
3          AGO       Angola  AG.LND.EL5M.ZS   
4          ALB      Albania  AG.LND.EL5M.ZS   

                           Series name SCALE Decimals      1990 1991 1992  \
0  Land area below 5m (% of land area)     0        1  29.57481   ..   ..   
1  Land area below 5m (% of land area)     0        1         0   ..   ..   
2  Land area below 5m (% of land area)     0        1         0   ..   ..   
3  Land area below 5m (% of land area)     0        1  0.208235   ..   ..   
4  Land area below 5m (% of land area)     0        1  4.967875   ..   ..   

  1993  ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
0   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
1   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
2   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
3   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
4   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  

[5 rows x 28 columns]

In [50]:

print("Descriptive statistics of the columns:")
data_orig.describe()

Descriptive statistics of the columns:


Country code Country name     Series code  \
count         13512        13512           13512   
unique          233          233              58   
top             ABW        Aruba  AG.LND.EL5M.ZS   
freq             58           58             233   

                                Series name  SCALE  Decimals   1990   1991  \
count                                 13512  13512     13512  10017  10017   
unique                                   58      2         3   4355   3398   
top     Land area below 5m (% of land area)      0         1     ..     ..   
freq                                    233  10017      5823   5163   6520   

         1992   1993  ...   2002   2003   2004   2005   2006   2007   2008  \
count   10017  10017  ...  10017  10017  10017  10017  10017  10017  10017   
unique   3523   3583  ...   3877   3869   4007   4484   4008   4047   4080   
top        ..     ..  ...     ..     ..     ..     ..     ..     ..     ..   
freq     6364   6300  ...   5960   5974   5792   4933   5781   5769   5414   

         2009   2010   2011  
count   10017  10017  12382  
unique   3506   2164   1434  
top        ..     ..     ..  
freq     6256   7685  10244  

[4 rows x 28 columns]


In order to better understand the nature of the columns "Series code", "Series name', "SCALE" and "Decimals", it is necessary to examine their values.

The following snippet prints the contents of the column 'Series name':

We printed Series name to understand the meaning of each row in the dataset so that we can work with the right data and avoid mistakes in analysis


In [54]:
data_orig['Series name'].unique()


array(['Land area below 5m (% of land area)',
       'Agricultural land under irrigation (% of total ag. land)',
       'Cereal yield (kg per hectare)',
       'Foreign direct investment, net inflows (% of GDP)',
       'Access to electricity (% of total population)',
       'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)',
       'Energy use per capita (kilograms of oil equivalent)',
       'CO2 emissions, total (KtCO2)',
       'CO2 emissions per capita (metric tons)',
       'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)',
       'Other GHG emissions, total (KtCO2e)',
       'Methane (CH4) emissions, total (KtCO2e)',
       'Nitrous oxide (N2O) emissions, total (KtCO2e)',
       'Annex-I emissions reduction target',
       'Disaster risk reduction progress score (1-5 scale; 5=best)',
       'GHG net emissions/removals by LUCF (MtCO2e)',
       'Hosted Clean Development Mechanism (CDM) projects',
       'Hosted Joint Implementation (JI) projects',
       'Av

Contents of the column 'Series code'

.unique() → This returns all the unique values in that column, removing duplicates.

In [65]:


data_orig['Series code'].unique()

array(['AG.LND.EL5M.ZS', 'AG.LND.IRIG.AG.ZS', 'AG.YLD.CREL.KG',
       'BX.KLT.DINV.WD.GD.ZS', 'EG.ELC.ACCS.ZS', 'EG.USE.COMM.GD.PP.KD',
       'EG.USE.PCAP.KG.OE', 'EN.ATM.CO2E.KT', 'EN.ATM.CO2E.PC',
       'EN.ATM.CO2E.PP.GD.KD', 'EN.ATM.GHGO.KT.CE', 'EN.ATM.METH.KT.CE',
       'EN.ATM.NOXE.KT.CE', 'EN.CLC.AERT', 'EN.CLC.DRSK.XQ',
       'EN.CLC.GHGR.MT.CE', 'EN.CLC.HCDM', 'EN.CLC.HJIP',
       'EN.CLC.HPPT.MM', 'EN.CLC.ICER', 'EN.CLC.IERU', 'EN.CLC.MDAT.ZS',
       'EN.CLC.MMDT.C', 'EN.CLC.NAMA', 'EN.CLC.NAPA', 'EN.CLC.NCOM',
       'EN.CLC.PCAT.C', 'EN.CLC.PCCC', 'EN.CLC.PCHW', 'EN.CLC.PCPT.MM',
       'EN.CLC.RNET', 'EN.POP.EL5M.ZS', 'EN.URB.MCTY.TL.ZS',
       'ER.H2O.FWTL.ZS', 'ER.LND.PTLD.ZS', 'IC.BUS.EASE.XQ',
       'IE.PPI.ENGY.CD', 'IE.PPI.TELE.CD', 'IE.PPI.TRAN.CD',
       'IE.PPI.WATR.CD', 'IQ.CPA.PUBS.XQ', 'IS.ROD.PAVE.ZS',
       'NY.GDP.MKTP.CD', 'NY.GNP.PCAP.CD', 'SE.ENR.PRSC.FM.ZS',
       'SE.PRM.CMPT.ZS', 'SH.DYN.MORT', 'SH.H2O.SAFE.ZS',
       'SH.MED.NUMW.P3', 'S


Contents of the column 'SCALE':

In [68]:
data_orig['SCALE'].unique()

array([0, 'Text'], dtype=object)


Contents of the column 'Decimals'

In [71]:

data_orig['Decimals'].unique()

array([1, 0, 'Text'], dtype=object)

At this point, it is unclear what are the values marked with the label 'Text' in the 'SCALE' and 'Decimals' columns. These are shown in the following tow outputs:

In [74]:
data_orig[data_orig['SCALE']=='Text']

Country code      Country name  Series code  \
3029          ABW             Aruba  EN.CLC.AERT   
3030          ADO           Andorra  EN.CLC.AERT   
3031          AFG       Afghanistan  EN.CLC.AERT   
3032          AGO            Angola  EN.CLC.AERT   
3033          ALB           Albania  EN.CLC.AERT   
...           ...               ...          ...   
7218          YEM       Yemen, Rep.  EN.CLC.RNET   
7219          ZAF      South Africa  EN.CLC.RNET   
7220          ZAR  Congo, Dem. Rep.  EN.CLC.RNET   
7221          ZMB            Zambia  EN.CLC.RNET   
7222          ZWE          Zimbabwe  EN.CLC.RNET   

                             Series name SCALE Decimals 1990 1991 1992 1993  \
3029  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3030  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3031  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3032  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3033  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
...                                  ...   ...      ...  ...  ...  ...  ...   
7218             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7219             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7220             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7221             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7222             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   

      ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
3029  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3030  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3031  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3032  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3033  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
7218  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7219  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  Yes  
7220  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7221  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7222  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  

[3495 rows x 28 columns]

data_orig[data_orig['Decimals']=='Text']

## Findings from the Global Overview

This global overview gives away the following facts about the available data:

- **Shape:** 28 columns, 13,512 rows  
- All columns are of type **"object"** – neither numeric nor pure string/text values  
- A certain amount of **missing values**, represented as both `NaN` (Not a Number) and the string `".."`  
- The rows marked as **'Text'** in the columns **'SCALE'** and **'Decimals'** contain almost no useful information (mostly NaN values)  
- The dataset columns represent **key values such as country**, corresponding **years**, and the **series code/name**  
- The columns **'Country name'**, **'Series code'**, **'SCALE'**, and **'Decimals'** do not provide any useful information and are therefore **obsolete**  
- The column **'Series name'** contains the **country-specific features** required for analysis  
- The names of the features in the column **'Series name'** are clear but **too long**  
- The dataset also contains **repeated country entries for different years**  
- **Data cleaning and transformation** will be required before performing any meaningful analysis or building ML models


***

## 3. Define the Initial Project Goals

The first overview of the raw data allows us to define **initial goals and objectives** of the machine learning project.  
These goals will be refined as more insights are gained from the data.  
This step helps in planning the **data cleaning, transformation, and visualization** strategy.

### Available Data Features (Country-Specific Parameters)

- **Greenhouse gas emissions** (CO₂, CH₄, N₂O, and others)  
- **Population parameters** (population count, urban population, population growth, etc.)  
- **Economic indicators** (GDP, GNI, Foreign Direct Investment, etc.)  
- **Land-related parameters** (cereal yield, agricultural land, protected areas, etc.)  
- **Climate data** (precipitation, natural disasters, etc.)  
- **Energy use**  
- **Health indicators** (counts of medical personnel, etc.)  

### Goal of the Project

- Investigate how **country-specific parameters** like economy, population, energy use, and land use affect **climate-related data** such as greenhouse gas emissions and precipitation.  
- Build a **machine learning model** that can **predict climate or emission data** based on other country-level features.  
- Continuously **refine these goals** as more insights are discovered during the data analysis phase.


***

## 4. Data cleaning

## Organization of the data cleaning and transformation
The main aim of the data cleaning and transformation is to represent the features (the country parameters contained in the column 'Series name') as separate columns and to make each row identifiable by a country and a year. At the same time, it would make sense to transform the years into a single column.

Additionally, it is necessary to get rid of empty rows or columns and deal with the remaining cells with missing values.

For these purposes, the following tasks have to be undertaken:

  1. Remove rows marked as "Text" in the "SCALE" and "Decimals" columns
  2. Remove the unnecessary columns "Country name", "Series code", "SCALE",      "Decimals"
  3. Transform the ".." strings and emplty cells ("") into NaN values for        easier recognission as missing values
  4. Transform all data columns into a numerical data type
  5. Rename the features in column "Series name"

 ### 4.1 Removing rows marked as "Text" in the "SCALE" and "Decimals" columns

In [100]:
# assign the data to a new DataFrame, which will be modified
data_clean = data_orig

print("Original number of rows:")
print(data_clean.shape[0])

# remove rows characterized as "Text" in the SCALE column
data_clean = data_clean[data_clean['SCALE']!='Text']

print("Current number of rows:")
print(data_clean.shape[0])

Original number of rows:
13512
Current number of rows:
10017


 ### 4.2 Removing the unnecessary columns "Country name", "Series code", "SCALE", "Decimals"

In [103]:
print("Original number of columns:")
print(data_clean.shape[1])

data_clean = data_clean.drop(['Country name', 'Series code', 'SCALE', 'Decimals'], axis='columns')

print("Current number of columns:")
print(data_clean.shape[1])

Original number of columns:
28
Current number of columns:
24


### 4.3 Transform the ".." strings and emplty cells ("") into NaN(not a number) values for easier recognission as missing values

In [125]:
data_clean.iloc[:,2:] = data_clean.iloc[:,2:].replace({'':np.nan, '..':np.nan})


# pandas will NOT automatically change column types (downcasting) in the background.
# This helps prevent unexpected data type changes, but the warning is only suppressed—actual type conversion should still be handled explicitly for best practice.
pd.set_option('future.no_silent_downcasting', True)



 ### 4.4 Transform all data columns into a numerical data type

In [136]:
# Apply pd.to_numeric to every column except the first two, converting text values to NaN where conversion fails.
for col in data_clean.columns[2:]:
    data_clean[col] = pd.to_numeric(data_clean[col], errors='coerce')

# Print the column data types after transformation
print("Print the column data types after transformation:")
print(data_clean.dtypes)


Print the column data types after transformation:
Country code     object
Series name      object
1990            float64
1991            float64
1992            float64
1993            float64
1994            float64
1995            float64
1996            float64
1997            float64
1998            float64
1999            float64
2000            float64
2001            float64
2002            float64
2003            float64
2004            float64
2005            float64
2006            float64
2007            float64
2008            float64
2009            float64
2010            float64
2011            float64
dtype: object



### 4.5 Rename the features in column "Series name"
The variable/feature names in the column 'Series name' are clear, but too long and not practical to use in the code. In order to improve that, the most relevant feature names will be renamed with shorter labels as indicated in the following table

| Variable name         | Description                                         | Unit                                   |
|-----------------------|-----------------------------------------------------|----------------------------------------|
| cereal_yield          | Cereal yield                                        | kg per hectare                         |
| fdi_perc_gdp          | Foreign direct investment, net inflows              | % of GDP                               |
| elec_access_perc      | Access to electricity                               | % of total population                  |
| en_per_gdp            | Energy use per units of GDP                         | kg oil eq./$1,000 of 2005 PPP $        |
| en_per_cap            | Energy use per capita                               | kilograms of oil equivalent            |
| co2_ttl               | CO2 emissions, total                                | KtCO2                                  |
| co2_per_cap           | CO2 emissions, total                                | metric tons                            |
| co2_per_gdp           | CO2 emissions per units of GDP                      | kg/$1,000 of 2005 PPP $                |
| other_ghg_ttl         | Other GHG emissions, total                          | KtCO2e                                 |
| ch4_ttl               | Methane (CH4) emissions, total                      | KtCO2                                  |
| n2o_ttl               | Nitrous oxide (N2O) emissions, total                | KtCO2                                  |
| nat_emerg             | Droughts, floods, extreme temps                     | % pop. avg. 1990-2009                  |
| pop_urb_aggl_perc     | Population in urban agglomerations >1million        | %                                      |
| prot_area_perc        | Nationally terrestrial protected areas              | % of total land area                   |
| gdp                   | Gross Domestic Product (GDP)                        | $                                      |
| gni_per_cap           | GNI per capita                                      | Atlas $                                |
| under_5_mort_rate     | Under-five mortality rate                           | per 1,000                              |
| pop_growth_perc       | Population growth                                   | annual %                               |
| pop                   | Population                                          | 1                                      |
| urb_pop_growth_perc   | Urban population growth                             | annual %                               |
| urb_pop               | Urban population                                    | 1                                      |


In [143]:
# define shorter names corresponding to most relevant variables in a dictionary
chosen_vars = {'Cereal yield (kg per hectare)': 'cereal_yield',
               'Foreign direct investment, net inflows (% of GDP)': 'fdi_perc_gdp',
               'Access to electricity (% of total population)': 'elec_access_perc',
               'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)': 'en_per_gdp',
               'Energy use per capita (kilograms of oil equivalent)': 'en_per_cap',
               'CO2 emissions, total (KtCO2)': 'co2_ttl',
               'CO2 emissions per capita (metric tons)': 'co2_per_cap',
               'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)': 'co2_per_gdp',
               'Other GHG emissions, total (KtCO2e)': 'other_ghg_ttl',
               'Methane (CH4) emissions, total (KtCO2e)': 'ch4_ttl',
               'Nitrous oxide (N2O) emissions, total (KtCO2e)': 'n2o_ttl',
               'Droughts, floods, extreme temps (% pop. avg. 1990-2009)': 'nat_emerg',
               'Population in urban agglomerations >1million (%)': 'pop_urb_aggl_perc',
               'Nationally terrestrial protected areas (% of total land area)': 'prot_area_perc',
               'GDP ($)': 'gdp',
               'GNI per capita (Atlas $)': 'gni_per_cap',
               'Under-five mortality rate (per 1,000)': 'under_5_mort_rate',
               'Population growth (annual %)': 'pop_growth_perc',
               'Population': 'pop',
               'Urban population growth (annual %)': 'urb_pop_growth_perc',
               'Urban population': 'urb_pop'
                }

# rename all variables in the column "Series name" with comprehensible shorter versions
data_clean2['Series name'] = data_clean2['Series name'].replace(to_replace=chosen_vars)

 ## 5. Data frame transformation
This is how the current data frame looks like

In [146]:
data_clean2.head()

Country code                          Series name       1990  1991  1992  \
0          ABW  Land area below 5m (% of land area)  29.574810   NaN   NaN   
1          ADO  Land area below 5m (% of land area)   0.000000   NaN   NaN   
2          AFG  Land area below 5m (% of land area)   0.000000   NaN   NaN   
3          AGO  Land area below 5m (% of land area)   0.208235   NaN   NaN   
4          ALB  Land area below 5m (% of land area)   4.967875   NaN   NaN   

   1993  1994  1995  1996  1997  ...  2002  2003  2004  2005  2006  2007  \
0   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
2   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
3   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
4   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   

   2008  2009  2010  2011  
0   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN  
3   NaN   NaN   NaN   NaN  
4   NaN   NaN   NaN   NaN  

[5 rows x 24 columns]

As mentioned earlier, we need to rearrange the data into a better format.
We’ll take the features listed in the ‘Series name’ column and make each one a separate column.
The years will go into one single column instead of being spread across many.
To do this, we pick the matching values for each feature along with their country and year, put them together into one column, and then combine everything into a new data frame — where all features line up correctly for the same country and year.

In [149]:
# save the short feature names into a list of strings
chosen_cols = list(chosen_vars.values())

# define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# iterate over all chosen features
for variable in chosen_cols:
    
    # pick only rows corresponding to the current feature
    frame = data_clean2[data_clean2['Series name'] == variable]
    
    # melt all the values for all years into one column and rename the columns correspondingly
    frame = frame.melt(id_vars=['Country code', 'Series name']).rename(columns={'Country code': 'country', 'variable': 'year', 'value': variable}).drop(['Series name'], axis='columns')
    
    # add the melted dataframe for the current feature into the list
    frame_list.append(frame)


# merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
from functools import reduce
all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country','year'], how='outer'), frame_list)

rearranging and combining data so that each feature (from the 'Series name' column) becomes its own column, and all the values for countries and years are neatly lined up in a single data frame.
After this transformation, the new data frame has the following layout:

In [152]:

all_vars.head()

country  year  cereal_yield  fdi_perc_gdp  elec_access_perc  en_per_gdp  \
0     ABW  1990           NaN           NaN               NaN         NaN   
1     ABW  1991           NaN     21.185138               NaN         NaN   
2     ABW  1992           NaN     -3.857809               NaN         NaN   
3     ABW  1993           NaN     -1.655492               NaN         NaN   
4     ABW  1994           NaN     -5.874439               NaN         NaN   

   en_per_cap   co2_ttl  co2_per_cap  co2_per_gdp  ...  nat_emerg  \
0         NaN  1840.834    29.620641          NaN  ...        NaN   
1         NaN  1928.842    29.838835          NaN  ...        NaN   
2         NaN  1723.490    25.239288          NaN  ...        NaN   
3         NaN  1771.161    24.405233          NaN  ...        NaN   
4         NaN  1763.827    22.977867          NaN  ...        NaN   

   pop_urb_aggl_perc  prot_area_perc           gdp  gni_per_cap  \
0                NaN        0.105547           NaN          NaN   
1                NaN        0.104930  8.720670e+08          NaN   
2                NaN        0.104930  9.586592e+08          NaN   
3                NaN        0.104930  1.083240e+09      14940.0   
4                NaN        0.104930  1.245810e+09      15320.0   

   under_5_mort_rate  pop_growth_perc      pop  urb_pop_growth_perc    urb_pop  
0                NaN         1.820254  62147.0             1.780501  31259.941  
1                NaN         3.936181  64642.0             3.337974  32321.000  
2                NaN         5.484041  68286.0             4.882234  33938.142  
3                NaN         6.088818  72573.0             5.483368  35851.062  
4                NaN         5.611677  76762.0             5.002538  37690.142  

[5 rows x 23 columns]

 ### 6. Remove the remaining missing values in an optimal way
 
Although some columns and rows with empty cells have already been deleted, there are still remaining missing values:

In [155]:
print("check the amount of missing values in each column")
all_vars.isnull().sum()

check the amount of missing values in each column


country                   0
year                      0
cereal_yield           1377
fdi_perc_gdp           1111
elec_access_perc       5027
en_per_gdp             2082
en_per_cap             1956
co2_ttl                1143
co2_per_cap            1146
co2_per_gdp            1557
other_ghg_ttl          4542
ch4_ttl                4526
n2o_ttl                4526
nat_emerg              4958
pop_urb_aggl_perc      2582
prot_area_perc          726
gdp                     779
gni_per_cap            1013
under_5_mort_rate       716
pop_growth_perc         278
pop                     252
urb_pop_growth_perc     490
urb_pop                 467
dtype: int64


When you clean a big dataset, you often find empty cells (called missing values or NaNs). You don't want to lose lots of useful information—so instead of deleting every row or column that has a missing value, you want to be smarter about what you remove.


- **Find which rows or columns have the most missing values.**
- **Start removing the ones with the most empty cells first.** This way, you get rid of the "emptiest" parts without throwing away lots of good data.
- **Count the number of missing cells (NaNs) for each country and each year** to see which years or countries are missing the most information.


- By removing the "sparsest" countries or years (those with the *most* missing data), you keep as many good columns and rows as possible.
- The goal is to end up with a table where most of the numbers are real, not empty.

**In short:** Instead of deleting everything with any missing value, you count which countries and years have the *most* gaps, and remove those first, so your cleaned dataset keeps as much good info as possible.

### 6.1 Filtering the years by missing values
Checking the amount of missing values for each year:
1. Make a note of each year in your data and start each with a count of 0   missing cells.

2. Go through each row of your table:
-  For each row, see which year it belongs to.
- In that row, count how many empty or missing cells (NaNs) there are.
- Add that number to the missing count for that year.
  
3. After checking all rows, sort the years from the one with the fewest     missing values to the one with the most.
4. Print out each year with the total number of missing values in that year

In [161]:
all_vars_clean = all_vars

#define an array with the unique year values
years_count_missing = dict.fromkeys(all_vars_clean['year'].unique(), 0)
for ind, row in all_vars_clean.iterrows():
    years_count_missing[row['year']] += row.isnull().sum()

# sort the years by missing values
years_missing_sorted = dict(sorted(years_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each year
print("missing values by year:")
for key, val in years_missing_sorted.items():
    print(key, ":", val)

missing values by year:
2005 : 1189
2000 : 1273
1995 : 1317
1990 : 1427
2007 : 1631
2006 : 1633
2004 : 1646
2008 : 1708
2003 : 1714
2002 : 1715
2001 : 1718
1999 : 1729
1998 : 1739
1997 : 1746
1996 : 1756
1994 : 1781
1993 : 1792
1992 : 1810
1991 : 1921
2009 : 2078
2010 : 3038
2011 : 4893


The purpose of the filtering is to delete rows with a significant amount of missing values for certain countries without removing too many years. So it is important to choose the proper limit for NaN values allowed per year. The previous output suggests to pick the years between 1991 and 2008 for the further analysis

In [167]:
print("number of missing values in the whole dataset before filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the years:")
print(all_vars_clean.shape[0])

# filter only rows for years between 1991 and 2008 (having less missing values)
all_vars_clean = all_vars_clean[(all_vars_clean['year'] >= 1991) & (all_vars_clean['year'] <= 2008)]

print("number of missing values in the whole dataset after filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the years:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the years:
41254
number of rows before filtering the years:
5126
number of missing values in the whole dataset after filtering the years:
29818
number of rows after filtering the years:
4194


### 6.2 Filtering the countries by missing values
The same procedure is applied to the filtering of countries with missing values. The following snippet shows the number of NaNs for each country.

In [172]:
# check the amount of missing values by country

# define an array with the unique country values
countries_count_missing = dict.fromkeys(all_vars_clean['country'].unique(), 0)

# iterate through all rows and count the amount of NaN values for each country
for ind, row in all_vars_clean.iterrows():
    countries_count_missing[row['country']] += row.isnull().sum()

# sort the countries by missing values
countries_missing_sorted = dict(sorted(countries_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each country
print("missing values by country:")
for key, val in countries_missing_sorted.items():
    print(key, ":", val)

missing values by country:
AGO : 81
ARG : 81
AUS : 81
AUT : 81
BGD : 81
BGR : 81
BOL : 81
BRA : 81
CAN : 81
CHE : 81
CHL : 81
CHN : 81
CIV : 81
CMR : 81
COG : 81
COL : 81
CRI : 81
DEU : 81
DNK : 81
DOM : 81
ECU : 81
EGY : 81
EMU : 81
ESP : 81
FIN : 81
FRA : 81
GBR : 81
GHA : 81
GTM : 81
HND : 81
HUN : 81
IDN : 81
IND : 81
IRL : 81
ISR : 81
ITA : 81
JOR : 81
JPN : 81
KEN : 81
KOR : 81
LAC : 81
LMC : 81
LMY : 81
MAR : 81
MEX : 81
MIC : 81
MNA : 81
MOZ : 81
MYS : 81
NGA : 81
NLD : 81
NZL : 81
PAK : 81
PAN : 81
PER : 81
PHL : 81
PRT : 81
PRY : 81
ROM : 81
SAS : 81
SAU : 81
SDN : 81
SEN : 81
SLV : 81
SWE : 81
SYR : 81
TGO : 81
THA : 81
TUR : 81
TZA : 81
UMC : 81
URY : 81
USA : 81
VEN : 81
VNM : 81
ZAF : 81
ZMB : 81
GRC : 82
POL : 82
YEM : 82
ZAR : 82
DZA : 84
ETH : 84
LIC : 84
SSA : 84
WLD : 84
ARE : 85
ECA : 85
RUS : 86
UKR : 86
ARM : 87
BLR : 87
UZB : 87
KAZ : 88
CZE : 89
IRN : 89
BEL : 90
AZE : 91
GEO : 92
LBN : 92
HTI : 94
NIC : 96
BEN : 99
BWA : 99
CYP : 99
GAB : 99
HIC : 99
JAM : 99
K

This output would suggest to remove rows for countries with more than 90 missing values:

In [180]:
print("number of missing values in the whole dataset before filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the countries:")
print(all_vars_clean.shape[0])


# filter only rows for countries with less than 90 missing values
countries_filter = []
for key, val in countries_missing_sorted.items():
    if val<90:
        countries_filter.append(key)

all_vars_clean = all_vars_clean[all_vars_clean['country'].isin(countries_filter)]

print("number of missing values in the whole dataset after filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the countries:")
print(all_vars_clean.shape[0])


number of missing values in the whole dataset before filtering the countries:
29818
number of rows before filtering the countries:
4194
number of missing values in the whole dataset after filtering the countries:
7854
number of rows after filtering the countries:
1728



### 6.3 Checking the features (columns) for missing values
The NaN values count in each column is:
- isnull() checks every cell in the DataFrame all_vars_clean and marks it as True if the value is missing (NaN), and False if it is present.

- .sum() then adds up all the True values in each column, because in Python True is treated as 1 and False as 0.

all_vars_clean.isnull().sum()

After having filtered the years and countries with the most missing values, the features elec_access_perc, other_ghg_ttl, ch4_ttl, n20_ttl and nat_emerg still contain a significant number of missing values, removing which will reduce the amount of observations very drastically. Therefore, the next step is to remove these columns:

In [188]:
# remove features with more than 20 missing values

from itertools import compress

# create a boolean mapping of features with more than 20 missing values
vars_bad = all_vars_clean.isnull().sum()>20

# remove the columns corresponding to the mapping of the features with many missing values
all_vars_clean2 = all_vars_clean.drop(compress(data = all_vars_clean.columns, selectors = vars_bad), axis='columns')

print("Remaining missing values per column:")
print(all_vars_clean2.isnull().sum())

Remaining missing values per column:
country                 0
year                    0
cereal_yield           10
fdi_perc_gdp           17
en_per_gdp              0
en_per_cap              0
co2_ttl                 9
co2_per_cap             9
co2_per_gdp             9
pop_urb_aggl_perc       0
prot_area_perc          0
gdp                     2
gni_per_cap            16
under_5_mort_rate       0
pop_growth_perc         0
pop                     0
urb_pop_growth_perc     0
urb_pop                 0
dtype: int64



Removing the rows with the remainin missing values will not impair the size of the dataset significantly, so these rows will be deleted:

In [191]:
# delete rows with any number of missing values
all_vars_clean3 = all_vars_clean2.dropna(axis='rows', how='any')

print("Remaining missing values per column:")
print(all_vars_clean3.isnull().sum())

print("Final shape of the cleaned dataset:")
print(all_vars_clean3.shape)

Remaining missing values per column:
country                0
year                   0
cereal_yield           0
fdi_perc_gdp           0
en_per_gdp             0
en_per_cap             0
co2_ttl                0
co2_per_cap            0
co2_per_gdp            0
pop_urb_aggl_perc      0
prot_area_perc         0
gdp                    0
gni_per_cap            0
under_5_mort_rate      0
pop_growth_perc        0
pop                    0
urb_pop_growth_perc    0
urb_pop                0
dtype: int64
Final shape of the cleaned dataset:
(1700, 18)


In [33]:
all_vars_clean2.head()

NameError: name 'all_vars_clean2' is not defined

 ## 7. Export of the cleaned data frame to a file
Now that the dataset has been rearranged and cleaned of missing values, it can be exported to a csv file (without the row index) for further analysis:

In [194]:
# export the clean dataframe to a csv file
all_vars_clean3.to_csv('data_cleaned.csv', index=False)

The further stages of the project are Data Visualization and Predictive Analysis, which can be found in the corresponding notebooks.

NameError: name 'all_vars_clean2' is not defined